### PR TITLE
Add unit tests for functions within `.~/reports/utils.js`

### PR DESCRIPTION
**Pull request by eason9487**
### Changes proposed in this Pull Request:

Add unit tests for functions within `.~/reports/utils.js`

- Copy test cases from woocommerce-admin repo for `getIdsFromQuery` function
- Add test cases for `aggregateIntervals`, `sumToPerformance`, and `addBaseToPerformance` functions
- Add a fallback `null` for `aggregateIntervals` function to make the returned invalid value consistent
- Remove the unneeded `missingFreeListingsData` fallback from `addBaseToPerformance` function

### Detailed test instructions:

1. Run `npm run test-unit` and make sure all these tests pass
2. The programs reports page should work well as before

### CODE REVIEW

File: js/src/reports/utils.js [hunk 1]
```diff
@@ -146,8 +146,7 @@ export function addBaseToPerformance( performance, base ) {
 			[ key ]: fieldsToPerformance(
 				performance[ key ].value,
 				base[ key ]?.value,
-				performance[ key ].missingFreeListingsData ||
-					base[ key ].missingFreeListingsData
```
Comment (eason9487) [inline js/src/reports/utils.js:150]: Remove this due to the `performance[ key ]` should always exist when entering the reduce by:

https://github.com/woocommerce/google-listings-and-ads/blob/648b1ebd31f0aa2244c5fd02c3bd186909b4b5f9/js/src/reports/utils.js#L142-L143

File: js/src/reports/utils.test.js [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+		} );
+		it( 'should ignore non numbers entries in the coma-separated list', () => {
+			expect( getIdsFromQuery( '77,,8,foo,null,9' ) ).toEqual( [
+				77,
+				8,
+				9,
+			] );
+		} );
+	} );
+} );
```
Comment (ecgan) [inline js/src/reports/utils.test.js:54]: ❓ I think we don't need to have this.

1. We implemented the `getIdsFromQuery` in PR https://github.com/woocommerce/google-listings-and-ads/pull/606 as a workaround for issue https://github.com/woocommerce/woocommerce-admin/issues/6980. If we wanted to write tests for it, we can write some simple ones for our version here; better tests should live in the upstream repo.
2. PR https://github.com/woocommerce/woocommerce-admin/pull/7028 has been merged, which means both `getIdsFromQuery` in WC Admin and our repo are now the same. We can just update our repo to use the updated package and remove our version of `getIdsFromQuery`, and we don't need test it in our repo here.

File: js/src/reports/utils.test.js [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+		const expectedIntervals = toIntervals(
+			[ '2021-01' ],
+			[ '2021-02' ],
+			[ '2021-03' ],
+			[ '2021-04' ],
+			[ '2021-05' ],
+			[ '2021-06' ],
+			[ '2021-07' ],
+			[ '2021-08' ]
+		);
```
Comment (ecgan) [inline js/src/reports/utils.test.js:108]: 💅  I would move this down a few lines (above `expectedIntervals.forEach`) so that it is closer to where it is being used.

File: js/src/reports/utils.test.js [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+		const result = aggregateIntervals( intervals1, intervals2 );
+
+		expectedIntervals.forEach( ( item, i ) => {
+			expect( result[ i ] ).toMatchObject( item );
+		} );
+	} );
+
+	describe( 'aggregate intervals', () => {
```
Comment (ecgan) [inline js/src/reports/utils.test.js]: ❓ The parent `describe` has the name `'aggregateIntervals'`, and this `describe` is named `'aggregate intervals'`, can we change the name here? Or perhaps we don't even need this `describe` here?

File: js/src/reports/utils.test.js [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+			( acc, value, i ) => ( {
+				...acc,
+				[ fields[ i ] ]: value,
+			} ),
+			{}
+		);
+	}
+
+	const toFreeTotals = toTotals.bind( null, freeFields );
+	const toPaidTotals = toTotals.bind( null, paidFields );
```
Comment (ecgan) [inline js/src/reports/utils.test.js]: ❓ Why do we want to use `.bind`? I feel like we don't need these two lines, and I'd prefer to write like `const paid = toTotals( paidFields, 123 );`. 🤔

File: js/src/reports/utils.test.js [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+	it( 'should be able to sum paid `totals` to performance only', () => {
+		const paid = toPaidTotals( 123 );
+		const free = undefined;
+		const performance = sumToPerformance( paid, free );
```
Comment (ecgan) [inline js/src/reports/utils.test.js:181]: 💅  I'd prefer to have an empty line above this, so that it is clearer that our tests is acting on this `sumToPerformance`.

File: js/src/reports/utils.test.js [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+		let result = addBaseToPerformance( performance, base );
+
+		expect( result ).toEqual( {} );
+
+		// Loaded
+		performance = { clicks: {}, sales: {} };
+		result = addBaseToPerformance( performance, base );
+
+		expect( result ).toHaveProperty( 'clicks' );
+		expect( result ).toHaveProperty( 'sales' );
```
Comment (ecgan) [inline js/src/reports/utils.test.js]: 💅  I think we should split this into a separate test, and also make the `performance` and `base` variable declaration follow the same order throughout all the tests. Since `performance` is the first parameter to `addBaseToPerformance`, we can declare `performance` first, similar to Line 271 https://github.com/woocommerce/google-listings-and-ads/pull/866/files#diff-23f7a4858f6f07bd5a18d44a21fe367623af0a07427c3c6e0fc049cb3b0a3c50R271.

**Review by ecgan — COMMENTED** [review]
Tested in my local machine, `npm run test-unit` all tests pass. Something is wrong in my local GLA site so I can't test the program reports page right now.

Left a few questions marked with ❓  ❔  and some 💅 .

File: js/src/reports/utils.test.js (hunk 3 above)
Comment (ecgan) [reply js/src/reports/utils.test.js:108]: I understand `expectedIntervals` is related to `intervals1` and `intervals2` so in a way it can make things easier to read by placing them together. We can live with this, don't worry too much about this. 😄

File: js/src/reports/utils.test.js (hunk 2 above)
Comment (tomalec) [reply js/src/reports/utils.test.js:54]: >  which means both `getIdsFromQuery` in WC Admin and our repo are now the same

I'm afraid not for all cases. 
1. It's still not released, https://github.com/woocommerce/woocommerce-admin/blob/main/packages/navigation/CHANGELOG.md. 
2. We lag quite far with our `/navigation` dependency https://github.com/woocommerce/google-listings-and-ads/blob/trunk/package.json#L63. 
3. What's even worse, we are tied to the version used in WC 5.2 https://github.com/woocommerce/google-listings-and-ads/blob/trunk/google-listings-and-ads.php#L13 what will lag us even longer.

-----

That's said I share the sentiment, of not duplicating the code too much. However, given the weight of technical debt our dependencies have, we will probably need to support this workaround for a quite long time - this would be nice to make sure the workaround works, without any regression.

Comment (tomalec) [reply js/src/reports/utils.test.js:54]: Maybe we could consider removing it from here, and creating a separate file, with a disclaimer/comment like:
```js
/** 
 * This is a test suite for a local copy, of not fully available fix in `@woocommerce/navigation`.
 * If you need to change something here, make sure you do so as well in the `/navigation` package as well.
 */

// Copied from https://github.com/woocommerce/woocommerce-admin/blob/b35156dcf17b44b3a81ea4b4528445b432917fd5/packages/navigation/src/test/index.js#L184-L221
```

File: js/src/reports/utils.test.js (hunk 5 above)
Comment (tomalec) [reply js/src/reports/utils.test.js]: ...but then you would have to write it every time (maybe we will use it more than 2 times in the future)
```js
toTotals( paidFields, 123 );
```
instead of
```js
toPaidTotals( 123 );
```

-----


Just saying :) Personally, I don't mind it either way

File: js/src/reports/utils.test.js [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+				[ fields[ i ] ]: value,
+			} ),
+			{}
+		);
+	}
+
+	const toFreeTotals = toTotals.bind( null, freeFields );
+	const toPaidTotals = toTotals.bind( null, paidFields );
+
+	it( 'should be able to sum paid `totals` to performance only', () => {
```
Comment (tomalec) [inline js/src/reports/utils.test.js]: :nail_care: I find this description quite vague. I don't know what to actually expect.
"be able" is ambiguous to me.

Is it like:
"When only paid totals are given, should map it to PerformanceData object"?

File: js/src/reports/utils.test.js [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+		expect( performance ).toMatchObject( {
+			[ paidFields[ 0 ] ]: {
+				value: 123,
+				delta: null,
+				missingFreeListingsData: MISSING_FREE_LISTINGS_DATA.NONE,
+			},
+		} );
+	} );
+
+	it( 'should be able to sum free `totals` to performance only', () => {
```
Comment (tomalec) [inline js/src/reports/utils.test.js]: Same doubts as in https://github.com/woocommerce/google-listings-and-ads/pull/866/files#r666510373

File: js/src/reports/utils.test.js [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+		expect( performance ).toMatchObject( {
+			[ sameFields[ 0 ] ]: {
+				value: 579,
+				delta: null,
+				missingFreeListingsData: MISSING_FREE_LISTINGS_DATA.NONE,
+			},
+		} );
+	} );
+
+	it( 'When a paid field is not (yet) available in API, should flag the data is not available', () => {
```
Comment (tomalec) [inline js/src/reports/utils.test.js]: :nail_care: This test and the one bellow, seems to be a special case of the one above? WDYT of grouping them in a `describe` so it could be easier to collapse it, and browse through the list of all specs.

For example, to something like:
```js
it( 'When paid and free `totals` are given', () => {
	it( 'should sum the same fields from paid and free `totals` to performance' );
	it( 'When a paid field is not (yet) available in API, should flag the data is not available' );
	it( `When an expected free field doesn't exist, should flag anticipated data is not returned from API` );
} );
```

**Comment by tomalec** [comment]
:scroll: :nail_care: If we aim for full coverage, we could add tests for aggregating intervals for a single one given https://github.com/woocommerce/google-listings-and-ads/blob/add/reports-utils-unit-test/js/src/reports/utils.js#L64

File: js/src/reports/utils.test.js (hunk 2 above)
Comment (eason9487) [reply js/src/reports/utils.test.js:54]: > 3\. What's even worse, we are tied to the version used in WC 5.2 https://github.com/woocommerce/google-listings-and-ads/blob/trunk/google-listings-and-ads.php#L13 what will lag us even longer.

For more information, the runtime `@woocommerce/navigation` version is bound to the version used in WC by:

https://github.com/woocommerce/google-listings-and-ads/blob/648b1ebd31f0aa2244c5fd02c3bd186909b4b5f9/webpack.config.js#L18-L30

For the longest case, we have to keep it until the minimum support WC version is bump to a version that includes the fix https://github.com/woocommerce/woocommerce-admin/pull/7028

File: js/src/reports/utils.test.js (hunk 4 above)
Comment (eason9487) [reply js/src/reports/utils.test.js]: Thanks! I didn't notice it after I adjusted the test groups several times during development. Removed by https://github.com/woocommerce/google-listings-and-ads/pull/866/commits/5410b5b3c0f5db06858a421995d5fd34cfead1bf

File: js/src/reports/utils.test.js (hunk 3 above)
Comment (eason9487) [reply js/src/reports/utils.test.js:108]: Adjusted by https://github.com/woocommerce/google-listings-and-ads/pull/866/commits/d55dc0405f9f550a6390cedb5fdcea5fa1312063

File: js/src/reports/utils.test.js (hunk 6 above)
Comment (eason9487) [reply js/src/reports/utils.test.js:181]: Adjusted by https://github.com/woocommerce/google-listings-and-ads/pull/866/commits/d55dc0405f9f550a6390cedb5fdcea5fa1312063

File: js/src/reports/utils.test.js (hunk 5 above)
Comment (eason9487) [reply js/src/reports/utils.test.js]: I originally used them many times, but then it became to only once each after adjusted the test cases several times. Since there are several other `toTotals` calls with different parameters that are not declared separately, I removed this usage by https://github.com/woocommerce/google-listings-and-ads/pull/866/commits/726b7d55d916c4c638fc5d9d08102fa45502b7ec

File: js/src/reports/utils.test.js (hunk 7 above)
Comment (eason9487) [reply js/src/reports/utils.test.js]: Separated by https://github.com/woocommerce/google-listings-and-ads/pull/866/commits/d303e33a142a67ea2124150e0815208fd302ca28

File: js/src/reports/utils.test.js (hunk 8 above)
Comment (eason9487) [reply js/src/reports/utils.test.js]: Updated by https://github.com/woocommerce/google-listings-and-ads/pull/866/commits/09ac902c6f757a67ff1d266588d6ba9387c6f8a2

File: js/src/reports/utils.test.js (hunk 9 above)
Comment (eason9487) [reply js/src/reports/utils.test.js]: Updated by https://github.com/woocommerce/google-listings-and-ads/pull/866/commits/09ac902c6f757a67ff1d266588d6ba9387c6f8a2

File: js/src/reports/utils.test.js (hunk 10 above)
Comment (eason9487) [reply js/src/reports/utils.test.js]: Added a group by https://github.com/woocommerce/google-listings-and-ads/pull/866/commits/09ac902c6f757a67ff1d266588d6ba9387c6f8a2

**Comment by eason9487** [comment]
> If we aim for full coverage, we could add tests for aggregating intervals for a single one given

Added by https://github.com/woocommerce/google-listings-and-ads/pull/866/commits/a5922bf58696e929dc1d0af5e36a5202e33403c2. And [a fallback was removed](https://github.com/woocommerce/google-listings-and-ads/commit/a5922bf58696e929dc1d0af5e36a5202e33403c2#diff-2bc2096e14fe5f43f0f1c25e0d050502f2f23a4f0283b092578f1cc7add0ce2fL122) because it will never get there due to the `freeTotals[ key ]` is checked ahead of that fallback:

https://github.com/woocommerce/google-listings-and-ads/blob/a5922bf58696e929dc1d0af5e36a5202e33403c2/js/src/reports/utils.js#L116-L119

File: js/src/reports/utils.test.js (hunk 2 above)
Comment (ecgan) [reply js/src/reports/utils.test.js:54]: That feels unfortunate. Gotcha, thank you all for the explanation. 🙂

**Comment by tomalec** [comment]
> because it will never get there due to the `freeTotals[ key ]` is checked ahead of that fallback

It is checked for `undefined` not for other falsy values, like `null` or `''`. AFAIR, I was guarding here against `null` as I've seen that in some responses. I'm not sure if that's still valid and reproducible.


something like:

```js
it( 'should sum the same fields from paid and free `totals` to performance, and consider `null` as `0`', () => {
	  const sameFields = [ 'ranks' ];
	  const paid = toTotals( sameFields, 123 );
	  const free = toTotals( sameFields, null );
	  
	  const performance = sumToPerformance( paid, free, sameFields );
	  
	  expect( performance ).toMatchObject( {
		  [ sameFields[ 0 ] ]: {
			  value: 123,
			  delta: null,
			  missingFreeListingsData: MISSING_FREE_LISTINGS_DATA.NONE,
		  },
	  } );
} );
```

**Comment by eason9487** [comment]
> It is checked for `undefined` not for other falsy values, like `null` or `''`. AFAIR, I was guarding here against `null` as I've seen that in some responses. I'm not sure if that's still valid and reproducible.

Thanks for remembering that. Before I removed the fallback, I traced both JS and PHP codes and thought the `freeTotals` parameter in `sumToPerformance` function should only appear the default empty object `{}` or an object with number type properties `{ clicks: 0, ...more fields }`. The `null` value probably would not happen.

https://github.com/woocommerce/google-listings-and-ads/blob/c63a721b9b29889bfcb4aa590e3261a6c75bb17a/src/API/Google/ReportTrait.php#L33-L42

https://github.com/woocommerce/google-listings-and-ads/blob/c63a721b9b29889bfcb4aa590e3261a6c75bb17a/js/src/reports/programs/useProgramsReport.js#L34-L38

---

But having further error-proofing is bringing more confidence. I added it back with a relevant test case by https://github.com/woocommerce/google-listings-and-ads/pull/866/commits/c63a721b9b29889bfcb4aa590e3261a6c75bb17a.

**Review by tomalec — APPROVED** [review]
LGTM :)

**Comment by tomalec** [comment]
:nail_care: 

> But having further error-proofing is bringing more confidence. I added it back with a relevant test case by c63a721.

On one hand, I like to be error-proof, but on the other, maybe it's not worth keeping such unnecessary code if you checked it should not happen. Maybe we could reduce the runtime code, and just add a note to code docs or explicit test to specify that only non-`null` values are expected.

**Comment by eason9487** [comment]
> On one hand, I like to be error-proof, but on the other, maybe it's not worth keeping such unnecessary code if you checked it should not happen. Maybe we could reduce the runtime code, and just add a note to code docs or explicit test to specify that only non-`null` values are expected.

Sounds good! Reverted and added note by https://github.com/woocommerce/google-listings-and-ads/pull/866/commits/de9ad3ee04986ca550198aa46bf59e3762b4706f, https://github.com/woocommerce/google-listings-and-ads/pull/866/commits/0aab2d5feaa1dbaf46e6ab4047d7469f1323f935
